[TESTING]: Failing PW tests

### ✅ Test Summary


Fix the following failing PW tests:

```
=========================== short test summary info ============================
FAILED tests/playwright/entities/test_prompts.py::TestPromptsCRUD::test_delete_prompt[chromium]
FAILED tests/playwright/entities/test_resources.py::TestResourcesCRUD::test_delete_resource[chromium]
FAILED tests/playwright/entities/test_resources_extended.py::TestResourcesViewModal::test_view_modal_different_resources_show_different_data[chromium]
FAILED tests/playwright/test_admin_menu_visibility.py::TestAdminMenuVisibility::test_platform_admin_sees_all_sections[chromium]
FAILED tests/playwright/test_admin_menu_visibility.py::TestAdminMenuVisibility::test_team_admin_sees_team_sections[chromium]
FAILED tests/playwright/test_admin_menu_visibility.py::TestAdminMenuVisibility::test_developer_sees_tool_sections[chromium]
FAILED tests/playwright/test_admin_menu_visibility.py::TestAdminMenuVisibility::test_viewer_sees_readonly_sections[chromium]
FAILED tests/playwright/test_admin_menu_visibility.py::TestAdminMenuVisibility::test_menu_sections_match_permissions[chromium]
FAILED tests/playwright/test_admin_url_context.py::TestAdminIframeContext::test_iframe_team_selector_onclick_stripped_but_delegation_works[chromium]
FAILED tests/playwright/test_organization.py::TestTeams::test_edit_settings_button[chromium]
FAILED tests/playwright/test_organization.py::TestTeams::test_edit_settings_max_members_defaults_to_global[chromium]
= 11 failed, 657 passed, 254 skipped, 8 warnings, 9 rerun in 2357.10s (0:39:17) =
❌ UI tests failed!
make: *** [Makefile:7122: test-ui-headless] Error 1
```

---

### 🧪 Test Type
Choose one or more applicable categories:

- [ ] Unit tests (Python)
- [ ] Integration / end-to-end tests
- [ ] Manual UI testing (admin panel)
- [ ] Transport coverage (HTTP, WebSocket, SSE, stdio)
- [ ] Federation / multi-gateway testing
- [ ] Configuration / environment-specific behavior
- [X] Other (explain below)

Playwright

---

### 🧬 Scope & Affected Components
Select what this test covers or validates:

- [ ] `mcpgateway` core (API logic, handlers)
- [X] Admin UI (HTMX / Alpine / Tailwind)
- [ ] Tool/Resource/Prompt logic
- [ ] Federation sync/discovery
- [ ] Server interactions or SSE
- [ ] Auth / JWT / security flows
- [ ] Observability (logging, metrics)
- [ ] Makefile, shell scripts or CLI
- [ ] Containerized setup (Docker/Podman)
- [ ] Other (explain below)

--- a/tests/AGENTS.md
+++ b/tests/AGENTS.md
@@ -139,12 +139,31 @@ All extend `BasePage` which provides `SidebarComponent` for tab navigation. Use 
 
 Fixtures in `conftest.py` handle login automatically (JWT cookie injection by default).
 
+### CSP-Safe JS Polling
+
+Strict CSP (`script-src 'self'`, no `unsafe-eval`) rejects `Page.wait_for_function()`/`Frame.wait_for_function()` immediately after navigation. Use `wait_for_js_condition()` instead — it polls via `evaluate()`, which isn't subject to that restriction:
+
+```python
+from .pages.admin_utils import wait_for_js_condition
+wait_for_js_condition(page, "typeof window.Admin !== 'undefined'", timeout=30000)
+```
+
 ### Writing New Playwright Tests
 
 1. Create a page object in `tests/playwright/pages/` extending `BasePage`
 2. Add a fixture in `tests/playwright/conftest.py`
 3. Create test file in `tests/playwright/test_<feature>.py`
 4. Use `pytest.skip()` when prerequisites aren't met (e.g., plugins not enabled)
+
+### Handling Async DB Operations
+
+After a DELETE, use `wait_for_entity_deleted()` to ride out DB commit propagation lag instead of asserting immediately:
+
+```python
+from .pages.admin_utils import wait_for_entity_deleted
+assert delete_resource(page, resource_id)
+assert wait_for_entity_deleted(page, "resources", resource_name)
+```
 
 ## JavaScript Unit Tests
 

--- a/tests/AGENTS.md
+++ b/tests/AGENTS.md
@@ -165,6 +165,22 @@ assert delete_resource(page, resource_id)
 assert wait_for_entity_deleted(page, "resources", resource_name)
 ```
 
+### Polling Utilities
+
+Lower-level primitives behind `wait_for_entity_deleted()`, for custom predicates:
+
+```python
+from .pages.admin_utils import wait_for_ui_by_retries, wait_for_ui_by_deadline
+
+# Fixed attempt count — API-poll style, each attempt cheap and uniform
+wait_for_ui_by_retries(page, lambda: check_something(), retries=5)
+
+# Wall-clock deadline — loops that also perform slow UI actions (navigation, reload)
+wait_for_ui_by_deadline(page, lambda: check_something(), deadline_seconds=8)
+```
+
+Both accept an optional `delay_ms(attempt) -> int` callable for backoff between attempts.
+
 ## JavaScript Unit Tests
 
 Admin UI JS logic is tested with Vitest (jsdom). Tests live in `tests/unit/js/`.

--- a/tests/playwright/conftest.py
+++ b/tests/playwright/conftest.py
@@ -27,6 +27,7 @@ from mcpgateway.config import Settings
 from tests.helpers.api_helpers import ApiTestHelper
 from tests.helpers.auth import make_test_jwt
 from .pages.admin_page import AdminPage
+from .pages.admin_utils import wait_for_js_condition
 from .pages.agents_page import AgentsPage
 from .pages.gateways_page import GatewaysPage
 from .pages.login_page import LoginPage
@@ -316,10 +317,13 @@ def _ensure_admin_logged_in(page: Page, base_url: str) -> None:
             raise AssertionError("Admin page failed to load: Internal Server Error (500)")
         raise
 
-    # Wait for JS initialization (showTab + HTMX + event delegation) before any tab clicks
+    # Wait for JS initialization (showTab + HTMX + event delegation) before any tab clicks.
+    # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by strict
+    # CSP (script-src 'self', no unsafe-eval) right after navigation.
     try:
-        page.wait_for_function(
-            "typeof window.Admin.showTab === 'function' && typeof window.htmx !== 'undefined' && window.Admin.eventDelegationInitialized === true",
+        wait_for_js_condition(
+            page,
+            "typeof window.Admin !== 'undefined' && typeof window.Admin.showTab === 'function' && typeof window.htmx !== 'undefined' && window.Admin.eventDelegationInitialized === true",
             timeout=30000,
         )
     except PlaywrightTimeoutError:

--- a/tests/playwright/conftest.py
+++ b/tests/playwright/conftest.py
@@ -318,8 +318,7 @@ def _ensure_admin_logged_in(page: Page, base_url: str) -> None:
         raise
 
     # Wait for JS initialization (showTab + HTMX + event delegation) before any tab clicks.
-    # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by strict
-    # CSP (script-src 'self', no unsafe-eval) right after navigation.
+    # See wait_for_js_condition() docstring for why evaluate()-based polling is used here.
     try:
         wait_for_js_condition(
             page,

--- a/tests/playwright/entities/test_resources.py
+++ b/tests/playwright/entities/test_resources.py
@@ -8,7 +8,7 @@ CRUD tests for Resources entity in ContextForge Admin UI.
 """
 
 # Local
-from ..pages.admin_utils import delete_resource, find_resource
+from ..pages.admin_utils import delete_resource, find_resource, wait_for_entity_deleted
 from ..pages.resources_page import ResourcesPage
 
 
@@ -50,5 +50,5 @@ class TestResourcesCRUD:
         # Delete using API helper
         assert delete_resource(resources_page.page, created_resource["id"])
 
-        # Verify deletion
-        assert find_resource(resources_page.page, test_resource_data["name"]) is None
+        # Verify deletion (retry to handle DB commit propagation lag)
+        assert wait_for_entity_deleted(resources_page.page, "resources", test_resource_data["name"]), f"Resource '{test_resource_data['name']}' still exists after deletion"

--- a/tests/playwright/pages/admin_utils.py
+++ b/tests/playwright/pages/admin_utils.py
@@ -10,15 +10,45 @@ Shared utility functions for admin page interactions.
 # Standard
 import logging
 import os
+import time
+from typing import Union
 import urllib.parse
 
 # Third-Party
-from playwright.sync_api import Page
+from playwright.sync_api import Frame, Page
 
 # Local
 from mcpgateway.admin import ADMIN_CSRF_COOKIE_NAME, ADMIN_CSRF_HEADER_NAME
 
 logger = logging.getLogger(__name__)
+
+
+def wait_for_js_condition(target: Union[Page, Frame], expression: str, timeout: int = 30000, polling: int = 100) -> None:
+    """Poll a boolean JS expression via ``evaluate()`` until it is truthy.
+
+    ``Page.wait_for_function``/``Frame.wait_for_function`` compile their predicate with
+    ``eval()``-equivalent semantics, which strict CSP (``script-src 'self'``, no
+    ``unsafe-eval``) rejects immediately with an ``EvalError`` right after navigation.
+    ``evaluate()`` is not subject to that restriction, so poll manually instead.
+
+    Args:
+        target: Page or Frame to evaluate the expression against.
+        expression: JS boolean expression to poll.
+        timeout: Maximum time to wait in milliseconds.
+        polling: Interval between polls in milliseconds.
+
+    Raises:
+        TimeoutError: If the expression never evaluates truthy within ``timeout``.
+    """
+    deadline = time.monotonic() + timeout / 1000
+    while time.monotonic() < deadline:
+        try:
+            if target.evaluate(expression):
+                return
+        except Exception:  # pylint: disable=broad-except
+            pass
+        target.wait_for_timeout(polling)
+    raise TimeoutError(f"Condition not met within {timeout}ms: {expression}")
 
 
 def _get_auth_headers(page: Page) -> dict:

--- a/tests/playwright/pages/admin_utils.py
+++ b/tests/playwright/pages/admin_utils.py
@@ -11,6 +11,7 @@ Shared utility functions for admin page interactions.
 import logging
 import os
 import time
+from typing import Callable, Optional
 import urllib.parse
 
 # Third-Party
@@ -49,10 +50,49 @@ def wait_for_js_condition(target: Page | Frame, expression: str, timeout: int = 
             last_exc = None
         except Exception as exc:  # pylint: disable=broad-except
             last_exc = exc
+            logger.debug("wait_for_js_condition: evaluate() raised while polling %r: %s", expression, exc)
         if time.monotonic() >= deadline:
             break
         target.wait_for_timeout(polling)
     raise PlaywrightTimeoutError(f"Condition not met within {timeout}ms: {expression}") from last_exc
+
+
+def wait_for_ui_condition(
+    page: Page,
+    predicate: Callable[[], bool],
+    retries: Optional[int] = None,
+    deadline_seconds: Optional[float] = None,
+    delay_ms: Optional[Callable[[int], int]] = None,
+) -> bool:
+    """Poll ``predicate()`` until truthy, bounded by attempt count or wall-clock deadline.
+
+    Give exactly one of ``retries``/``deadline_seconds`` as the stop condition — attempt
+    count for API-poll style retries, wall-clock deadline for loops that also perform
+    slow UI actions (navigation, reload) between checks.
+
+    Args:
+        page: Playwright page, used for ``wait_for_timeout`` between attempts.
+        predicate: Zero-arg callable returning a truthy value once polling should stop.
+        retries: Maximum number of attempts.
+        deadline_seconds: Maximum wall-clock time to keep polling, in seconds.
+        delay_ms: Optional callable mapping the 0-indexed attempt number to a backoff
+            delay in milliseconds, applied before the next attempt.
+
+    Returns:
+        True if predicate() returned truthy before exhausting retries/deadline.
+    """
+    deadline = time.monotonic() + deadline_seconds if deadline_seconds is not None else None
+    attempt = 0
+    while True:
+        if predicate():
+            return True
+        if deadline is not None and time.monotonic() >= deadline:
+            return False
+        if retries is not None and attempt + 1 >= retries:
+            return False
+        if delay_ms is not None:
+            page.wait_for_timeout(delay_ms(attempt))
+        attempt += 1
 
 
 def _get_auth_headers(page: Page) -> dict:
@@ -141,16 +181,19 @@ def wait_for_entity_deleted(page: Page, endpoint: str, name: str, retries: int =
         True if entity disappeared within retries, False otherwise.
     """
     headers = _get_auth_headers(page)
-    for attempt in range(retries):
-        url = f"/admin/{endpoint}?per_page=500&cache_bust=del{attempt}"
+
+    def _not_found() -> bool:
+        url = f"/admin/{endpoint}?per_page=500&cache_bust=del{_not_found.calls}"
+        _not_found.calls += 1
         response = page.request.get(url, headers=headers)
-        if response.ok:
-            payload = response.json()
-            data = payload if isinstance(payload, list) else payload.get("data", [])
-            if not any(item.get("name") == name for item in data):
-                return True
-        page.wait_for_timeout(min(100 * (2**attempt), 800))
-    return False
+        if not response.ok:
+            return False
+        payload = response.json()
+        data = payload if isinstance(payload, list) else payload.get("data", [])
+        return not any(item.get("name") == name for item in data)
+
+    _not_found.calls = 0
+    return wait_for_ui_condition(page, _not_found, retries=retries, delay_ms=lambda attempt: min(100 * (2**attempt), 800))
 
 
 def find_tool(page: Page, tool_name: str, retries: int = 5):

--- a/tests/playwright/pages/admin_utils.py
+++ b/tests/playwright/pages/admin_utils.py
@@ -11,11 +11,11 @@ Shared utility functions for admin page interactions.
 import logging
 import os
 import time
-from typing import Union
 import urllib.parse
 
 # Third-Party
 from playwright.sync_api import Frame, Page
+from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
 
 # Local
 from mcpgateway.admin import ADMIN_CSRF_COOKIE_NAME, ADMIN_CSRF_HEADER_NAME
@@ -23,7 +23,7 @@ from mcpgateway.admin import ADMIN_CSRF_COOKIE_NAME, ADMIN_CSRF_HEADER_NAME
 logger = logging.getLogger(__name__)
 
 
-def wait_for_js_condition(target: Union[Page, Frame], expression: str, timeout: int = 30000, polling: int = 100) -> None:
+def wait_for_js_condition(target: Page | Frame, expression: str, timeout: int = 30000, polling: int = 100) -> None:
     """Poll a boolean JS expression via ``evaluate()`` until it is truthy.
 
     ``Page.wait_for_function``/``Frame.wait_for_function`` compile their predicate with
@@ -38,17 +38,21 @@ def wait_for_js_condition(target: Union[Page, Frame], expression: str, timeout: 
         polling: Interval between polls in milliseconds.
 
     Raises:
-        TimeoutError: If the expression never evaluates truthy within ``timeout``.
+        PlaywrightTimeoutError: If the expression never evaluates truthy within ``timeout``.
     """
     deadline = time.monotonic() + timeout / 1000
-    while time.monotonic() < deadline:
+    last_exc: Exception | None = None
+    while True:
         try:
             if target.evaluate(expression):
                 return
-        except Exception:  # pylint: disable=broad-except
-            pass
+            last_exc = None
+        except Exception as exc:  # pylint: disable=broad-except
+            last_exc = exc
+        if time.monotonic() >= deadline:
+            break
         target.wait_for_timeout(polling)
-    raise TimeoutError(f"Condition not met within {timeout}ms: {expression}")
+    raise PlaywrightTimeoutError(f"Condition not met within {timeout}ms: {expression}") from last_exc
 
 
 def _get_auth_headers(page: Page) -> dict:

--- a/tests/playwright/pages/admin_utils.py
+++ b/tests/playwright/pages/admin_utils.py
@@ -11,7 +11,7 @@ Shared utility functions for admin page interactions.
 import logging
 import os
 import time
-from typing import Callable, Optional
+from typing import Callable
 import urllib.parse
 
 # Third-Party
@@ -57,38 +57,63 @@ def wait_for_js_condition(target: Page | Frame, expression: str, timeout: int = 
     raise PlaywrightTimeoutError(f"Condition not met within {timeout}ms: {expression}") from last_exc
 
 
-def wait_for_ui_condition(
+def wait_for_ui_by_retries(
     page: Page,
     predicate: Callable[[], bool],
-    retries: Optional[int] = None,
-    deadline_seconds: Optional[float] = None,
-    delay_ms: Optional[Callable[[int], int]] = None,
+    retries: int,
+    delay_ms: Callable[[int], int] | None = None,
 ) -> bool:
-    """Poll ``predicate()`` until truthy, bounded by attempt count or wall-clock deadline.
+    """Poll ``predicate()`` until truthy, bounded by a fixed attempt count.
 
-    Give exactly one of ``retries``/``deadline_seconds`` as the stop condition — attempt
-    count for API-poll style retries, wall-clock deadline for loops that also perform
-    slow UI actions (navigation, reload) between checks.
+    Use for API-poll style retries where each attempt is cheap and uniform.
 
     Args:
         page: Playwright page, used for ``wait_for_timeout`` between attempts.
         predicate: Zero-arg callable returning a truthy value once polling should stop.
         retries: Maximum number of attempts.
+        delay_ms: Optional callable mapping the 0-indexed attempt number to a backoff
+            delay in milliseconds, applied before the next attempt.
+
+    Returns:
+        True if predicate() returned truthy before exhausting retries.
+    """
+    for attempt in range(retries):
+        if predicate():
+            return True
+        if attempt + 1 >= retries:
+            return False
+        if delay_ms is not None:
+            page.wait_for_timeout(delay_ms(attempt))
+    return False
+
+
+def wait_for_ui_by_deadline(
+    page: Page,
+    predicate: Callable[[], bool],
+    deadline_seconds: float,
+    delay_ms: Callable[[int], int] | None = None,
+) -> bool:
+    """Poll ``predicate()`` until truthy, bounded by a wall-clock deadline.
+
+    Use for loops that also perform slow UI actions (navigation, reload) between
+    checks, where attempt count isn't a meaningful bound.
+
+    Args:
+        page: Playwright page, used for ``wait_for_timeout`` between attempts.
+        predicate: Zero-arg callable returning a truthy value once polling should stop.
         deadline_seconds: Maximum wall-clock time to keep polling, in seconds.
         delay_ms: Optional callable mapping the 0-indexed attempt number to a backoff
             delay in milliseconds, applied before the next attempt.
 
     Returns:
-        True if predicate() returned truthy before exhausting retries/deadline.
+        True if predicate() returned truthy before the deadline elapsed.
     """
-    deadline = time.monotonic() + deadline_seconds if deadline_seconds is not None else None
+    deadline = time.monotonic() + deadline_seconds
     attempt = 0
     while True:
         if predicate():
             return True
-        if deadline is not None and time.monotonic() >= deadline:
-            return False
-        if retries is not None and attempt + 1 >= retries:
+        if time.monotonic() >= deadline:
             return False
         if delay_ms is not None:
             page.wait_for_timeout(delay_ms(attempt))
@@ -193,7 +218,7 @@ def wait_for_entity_deleted(page: Page, endpoint: str, name: str, retries: int =
         return not any(item.get("name") == name for item in data)
 
     _not_found.calls = 0
-    return wait_for_ui_condition(page, _not_found, retries=retries, delay_ms=lambda attempt: min(100 * (2**attempt), 800))
+    return wait_for_ui_by_retries(page, _not_found, retries=retries, delay_ms=lambda attempt: min(100 * (2**attempt), 800))
 
 
 def find_tool(page: Page, tool_name: str, retries: int = 5):

--- a/tests/playwright/security/test_innerhtml_guard_iframe.py
+++ b/tests/playwright/security/test_innerhtml_guard_iframe.py
@@ -37,7 +37,7 @@ from __future__ import annotations
 from typing import Any, Dict
 
 # Third-Party
-from playwright.sync_api import FrameLocator, Page
+from playwright.sync_api import Frame, FrameLocator, Page
 from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
 import pytest
 
@@ -49,7 +49,7 @@ from ..pages.admin_utils import wait_for_js_condition
 # ---------------------------------------------------------------------------
 
 
-def _get_admin_frame(page: Page):
+def _get_admin_frame(page: Page) -> Frame:
     """Return the Frame object for the admin iframe.
 
     ``iframe_host`` yields a ``FrameLocator``.  To call ``evaluate()`` we
@@ -171,11 +171,12 @@ class TestTeamSelectorInIframe:
                 const btn = container.querySelector('.team-selector-item');
                 if (!btn) return { skipped: true, reason: 'button not found' };
 
-                // Spy on selectTeamFromSelector
+                // Spy on selectTeamFromSelector (exposed as window.Admin.selectTeamFromSelector,
+                // which is what eventDelegation.js's click handler dispatches to)
                 let called = false;
                 let calledWith = null;
-                const orig = window.selectTeamFromSelector;
-                window.selectTeamFromSelector = (el) => {
+                const orig = window.Admin.selectTeamFromSelector;
+                window.Admin.selectTeamFromSelector = (el) => {
                     called = true;
                     calledWith = el ? el.getAttribute('data-team-id') : null;
                 };
@@ -184,7 +185,7 @@ class TestTeamSelectorInIframe:
                     btn.click();
                     return { skipped: false, called, calledWith };
                 } finally {
-                    window.selectTeamFromSelector = orig;
+                    window.Admin.selectTeamFromSelector = orig;
                     container.innerHTML = '';
                 }
             }

--- a/tests/playwright/security/test_innerhtml_guard_iframe.py
+++ b/tests/playwright/security/test_innerhtml_guard_iframe.py
@@ -41,6 +41,9 @@ from playwright.sync_api import FrameLocator, Page
 from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
 import pytest
 
+# Local
+from ..pages.admin_utils import wait_for_js_condition
+
 # ---------------------------------------------------------------------------
 # Helper: get the underlying Frame object from the iframe_host FrameLocator
 # ---------------------------------------------------------------------------
@@ -53,10 +56,20 @@ def _get_admin_frame(page: Page):
     need the underlying ``Frame`` object, which Playwright exposes via
     ``page.frames``.  The admin iframe is the first child frame of the host
     page (index 1; index 0 is the host page itself).
+
+    ``page.main_frame`` must always be skipped: the host page is built via
+    ``page.set_content()``, which replaces the DOM in place without a real
+    navigation, so ``page.main_frame.url`` keeps reporting whatever URL the
+    page had *before* ``set_content()`` ran (typically the real ``/admin/``
+    URL from the login flow) even though its ``window.Admin`` now belongs to
+    a torn-down document. Matching on that stale URL string would silently
+    hand callers the host frame instead of the actual embedded admin iframe.
     """
     # Wait until the iframe has navigated to /admin/
     for _ in range(30):
         for frame in page.frames:
+            if frame is page.main_frame:
+                continue
             if "/admin" in frame.url:
                 return frame
         page.wait_for_timeout(500)
@@ -357,6 +370,11 @@ class TestMetricsRetryInIframe:
         """showMetricsError() retry button must have data-action, not onclick, inside iframe."""
         frame = _get_admin_frame(page)
 
+        # metrics.js is lazy-loaded on demand; Admin.showMetricsError only becomes
+        # callable once the metrics tab has triggered its chunk load.
+        iframe_host.locator("#tab-metrics").click()
+        wait_for_js_condition(frame, "typeof window.Admin.showMetricsError === 'function'", timeout=15000)
+
         result: Dict[str, Any] = frame.evaluate("""
             () => {
                 let container = document.getElementById('aggregated-metrics-content');
@@ -390,6 +408,11 @@ class TestMetricsRetryInIframe:
     def test_metrics_error_retry_button_calls_retryLoadMetrics_inside_iframe(self, iframe_host: FrameLocator, page: Page) -> None:
         """Clicking the retry button inside the iframe must trigger retry behavior."""
         frame = _get_admin_frame(page)
+
+        # metrics.js is lazy-loaded on demand; Admin.showMetricsError only becomes
+        # callable once the metrics tab has triggered its chunk load.
+        iframe_host.locator("#tab-metrics").click()
+        wait_for_js_condition(frame, "typeof window.Admin.showMetricsError === 'function'", timeout=15000)
 
         result: Dict[str, Any] = frame.evaluate("""
             () => {
@@ -425,6 +448,11 @@ class TestMetricsRetryInIframe:
     def test_metrics_empty_state_refresh_button_has_data_action_inside_iframe(self, iframe_host: FrameLocator, page: Page) -> None:
         """displayMetrics({}) empty-state button must have data-action, not onclick, inside iframe."""
         frame = _get_admin_frame(page)
+
+        # metrics.js is lazy-loaded on demand; Admin.displayMetrics only becomes
+        # callable once the metrics tab has triggered its chunk load.
+        iframe_host.locator("#tab-metrics").click()
+        wait_for_js_condition(frame, "typeof window.Admin.displayMetrics === 'function'", timeout=15000)
 
         result: Dict[str, Any] = frame.evaluate("""
             () => {

--- a/tests/playwright/security/test_token_lifecycle.py
+++ b/tests/playwright/security/test_token_lifecycle.py
@@ -21,7 +21,8 @@ import time
 import uuid
 
 # Third-Party
-from playwright.sync_api import APIRequestContext
+from playwright.sync_api import APIRequestContext, expect
+from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
 import pytest
 
 logger = logging.getLogger(__name__)
@@ -189,19 +190,27 @@ class TestTokenRevokeUI:
         revoke_btn_count = None
         token_card_count = None
         while time.monotonic() < deadline:
+            tokens_page.page.reload(wait_until="domcontentloaded")
             tokens_page.navigate_to_tokens_tab()
             # Enable "Show inactive" to make revoked tokens visible in the list
             inactive_checkbox = tokens_page.page.locator("#tokens-inactive-toggle, [name='include_inactive']")
             if inactive_checkbox.count() > 0 and not inactive_checkbox.first.is_checked():
                 inactive_checkbox.first.click()
-                tokens_page.page.wait_for_timeout(500)
 
-            token_card_count = tokens_page.page.locator(f"text={token_name}").count()
+            # Wait for the list to actually settle before reading counts. Without this,
+            # a not-yet-rendered list (0 cards) is indistinguishable from a genuinely
+            # absent token, and the loop below would break on the first pass and fall
+            # through to the API-only check — skipping the UI assertion entirely.
+            token_card = tokens_page.page.locator(f"text={token_name}")
+            try:
+                expect(token_card.first).to_be_visible(timeout=3000)
+            except (PlaywrightTimeoutError, AssertionError):
+                pass
+
+            token_card_count = token_card.count()
             revoke_btn_count = tokens_page.get_token_revoke_btn(token_name).count()
             if token_card_count == 0 or revoke_btn_count == 0:
                 break
-
-            tokens_page.page.reload(wait_until="domcontentloaded")
 
         # Verify the token card IS rendered (not vacuously absent)
         if token_card_count and token_card_count > 0:

--- a/tests/playwright/security/test_token_lifecycle.py
+++ b/tests/playwright/security/test_token_lifecycle.py
@@ -17,6 +17,7 @@ from __future__ import annotations
 
 # Standard
 import logging
+import time
 import uuid
 
 # Third-Party
@@ -179,20 +180,33 @@ class TestTokenRevokeUI:
         revoke_resp = admin_api.delete(f"/tokens/{token_id}")
         assert revoke_resp.status in (200, 204)
 
-        # Navigate to tokens tab with include_inactive=true so revoked token IS listed
-        tokens_page.navigate_to_tokens_tab()
-        # Enable "Show inactive" to make revoked tokens visible in the list
-        inactive_checkbox = tokens_page.page.locator("#tokens-inactive-toggle, [name='include_inactive']")
-        if inactive_checkbox.count() > 0 and not inactive_checkbox.first.is_checked():
-            inactive_checkbox.first.click()
-            tokens_page.page.wait_for_timeout(2000)
+        # Navigate to tokens tab with include_inactive=true so revoked token IS listed.
+        # The nginx admin-page cache has a 5s TTL keyed on URL+cookies (see
+        # infra/nginx/nginx.conf), so a GET for this exact listing issued by an
+        # earlier test in the same session can still be served stale here. Retry
+        # past that window instead of asserting on a single fetch.
+        deadline = time.monotonic() + 8
+        revoke_btn_count = None
+        token_card_count = None
+        while time.monotonic() < deadline:
+            tokens_page.navigate_to_tokens_tab()
+            # Enable "Show inactive" to make revoked tokens visible in the list
+            inactive_checkbox = tokens_page.page.locator("#tokens-inactive-toggle, [name='include_inactive']")
+            if inactive_checkbox.count() > 0 and not inactive_checkbox.first.is_checked():
+                inactive_checkbox.first.click()
+                tokens_page.page.wait_for_timeout(500)
+
+            token_card_count = tokens_page.page.locator(f"text={token_name}").count()
+            revoke_btn_count = tokens_page.get_token_revoke_btn(token_name).count()
+            if token_card_count == 0 or revoke_btn_count == 0:
+                break
+
+            tokens_page.page.reload(wait_until="domcontentloaded")
 
         # Verify the token card IS rendered (not vacuously absent)
-        token_card = tokens_page.page.locator(f"text={token_name}")
-        if token_card.count() > 0:
+        if token_card_count and token_card_count > 0:
             # Token is visible — the revoke button must NOT be present
-            revoke_btn = tokens_page.get_token_revoke_btn(token_name)
-            assert revoke_btn.count() == 0, "Revoke button should be hidden for already-revoked tokens"
+            assert revoke_btn_count == 0, "Revoke button should be hidden for already-revoked tokens"
         else:
             # If include_inactive toggle is not available, verify via API that token is truly revoked
             resp = admin_api.get(f"/tokens/{token_id}")

--- a/tests/playwright/security/test_token_lifecycle.py
+++ b/tests/playwright/security/test_token_lifecycle.py
@@ -17,13 +17,15 @@ from __future__ import annotations
 
 # Standard
 import logging
-import time
 import uuid
 
 # Third-Party
 from playwright.sync_api import APIRequestContext, expect
 from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
 import pytest
+
+# Local
+from ..pages.admin_utils import wait_for_ui_condition
 
 logger = logging.getLogger(__name__)
 
@@ -182,18 +184,13 @@ class TestTokenRevokeUI:
         assert revoke_resp.status in (200, 204)
 
         # Navigate to tokens tab with include_inactive=true so revoked token IS listed.
-        # The nginx admin-page cache has a 5s TTL keyed on URL+cookies (see
-        # infra/nginx/nginx.conf), so a GET for this exact listing issued by an
-        # earlier test in the same session can still be served stale here. Retry
-        # past that window instead of asserting on a single fetch.
-        deadline = time.monotonic() + 8
-        revoke_btn_count = None
-        token_card_count = None
-        while time.monotonic() < deadline:
+        counts = {"card": None, "revoke": None}
+
+        def _settled() -> bool:
             tokens_page.page.reload(wait_until="domcontentloaded")
             tokens_page.navigate_to_tokens_tab()
             # Enable "Show inactive" to make revoked tokens visible in the list
-            inactive_checkbox = tokens_page.page.locator("#tokens-inactive-toggle, [name='include_inactive']")
+            inactive_checkbox = tokens_page.page.locator("#show-inactive-tokens")
             if inactive_checkbox.count() > 0 and not inactive_checkbox.first.is_checked():
                 inactive_checkbox.first.click()
 
@@ -207,10 +204,17 @@ class TestTokenRevokeUI:
             except (PlaywrightTimeoutError, AssertionError):
                 pass
 
-            token_card_count = token_card.count()
-            revoke_btn_count = tokens_page.get_token_revoke_btn(token_name).count()
-            if token_card_count == 0 or revoke_btn_count == 0:
-                break
+            counts["card"] = token_card.count()
+            counts["revoke"] = tokens_page.get_token_revoke_btn(token_name).count()
+            return counts["card"] == 0 or counts["revoke"] == 0
+
+        # The nginx admin-page cache has a 5s TTL keyed on URL+cookies (see
+        # infra/nginx/nginx.conf), so a GET for this exact listing issued by an
+        # earlier test in the same session can still be served stale here. Retry
+        # past that window instead of asserting on a single fetch.
+        wait_for_ui_condition(tokens_page.page, _settled, deadline_seconds=8)
+        token_card_count = counts["card"]
+        revoke_btn_count = counts["revoke"]
 
         # Verify the token card IS rendered (not vacuously absent)
         if token_card_count and token_card_count > 0:

--- a/tests/playwright/security/test_token_lifecycle.py
+++ b/tests/playwright/security/test_token_lifecycle.py
@@ -25,7 +25,7 @@ from playwright.sync_api import TimeoutError as PlaywrightTimeoutError
 import pytest
 
 # Local
-from ..pages.admin_utils import wait_for_ui_condition
+from ..pages.admin_utils import wait_for_ui_by_deadline
 
 logger = logging.getLogger(__name__)
 
@@ -212,7 +212,7 @@ class TestTokenRevokeUI:
         # infra/nginx/nginx.conf), so a GET for this exact listing issued by an
         # earlier test in the same session can still be served stale here. Retry
         # past that window instead of asserting on a single fetch.
-        wait_for_ui_condition(tokens_page.page, _settled, deadline_seconds=8)
+        wait_for_ui_by_deadline(tokens_page.page, _settled, deadline_seconds=8)
         token_card_count = counts["card"]
         revoke_btn_count = counts["revoke"]
 

--- a/tests/playwright/test_admin_menu_visibility.py
+++ b/tests/playwright/test_admin_menu_visibility.py
@@ -155,10 +155,9 @@ def _wait_for_admin_shell(page: Page, timeout: int = 60000) -> None:
         if "Internal Server Error" in content:
             raise AssertionError("Admin page failed to load: Internal Server Error (500)")
         raise
-    # Wait for JS initialization. Uses evaluate()-based polling rather than
-    # wait_for_function(), whose eval() mechanism strict CSP (script-src 'self',
-    # no unsafe-eval) rejects right after navigation. showTab lives on the
-    # Admin namespace, not as a bare global, since the JS bundling refactor.
+    # Wait for JS initialization (see wait_for_js_condition() docstring for why
+    # evaluate()-based polling is used). showTab lives on the Admin namespace,
+    # not as a bare global, since the JS bundling refactor.
     try:
         wait_for_js_condition(
             page,

--- a/tests/playwright/test_admin_menu_visibility.py
+++ b/tests/playwright/test_admin_menu_visibility.py
@@ -32,6 +32,7 @@ import pytest
 from tests.helpers.api_helpers import ApiTestHelper
 from tests.helpers.auth import make_playwright_api_context, make_test_jwt
 from .conftest import BASE_URL
+from .pages.admin_utils import wait_for_js_condition
 
 logger = logging.getLogger(__name__)
 
@@ -154,13 +155,17 @@ def _wait_for_admin_shell(page: Page, timeout: int = 60000) -> None:
         if "Internal Server Error" in content:
             raise AssertionError("Admin page failed to load: Internal Server Error (500)")
         raise
-    # Wait for JS initialization
+    # Wait for JS initialization. Uses evaluate()-based polling rather than
+    # wait_for_function(), whose eval() mechanism strict CSP (script-src 'self',
+    # no unsafe-eval) rejects right after navigation. showTab lives on the
+    # Admin namespace, not as a bare global, since the JS bundling refactor.
     try:
-        page.wait_for_function(
-            "typeof window.showTab === 'function' && typeof window.htmx !== 'undefined'",
+        wait_for_js_condition(
+            page,
+            "typeof window.Admin !== 'undefined' && typeof window.Admin.showTab === 'function' && typeof window.htmx !== 'undefined'",
             timeout=30000,
         )
-    except PlaywrightTimeoutError:
+    except TimeoutError:
         pass
 
 

--- a/tests/playwright/test_admin_menu_visibility.py
+++ b/tests/playwright/test_admin_menu_visibility.py
@@ -165,7 +165,7 @@ def _wait_for_admin_shell(page: Page, timeout: int = 60000) -> None:
             "typeof window.Admin !== 'undefined' && typeof window.Admin.showTab === 'function' && typeof window.htmx !== 'undefined'",
             timeout=30000,
         )
-    except TimeoutError:
+    except PlaywrightTimeoutError:
         pass
 
 

--- a/tests/playwright/test_admin_url_context.py
+++ b/tests/playwright/test_admin_url_context.py
@@ -26,6 +26,7 @@ import pytest
 
 # Local
 from .conftest import _ensure_admin_logged_in
+from .pages.admin_utils import wait_for_js_condition
 
 # A placeholder team_id value; tests use it as a URL param and verify it survives
 # mutations.  In a real team-scoped deployment this would be a valid UUID.
@@ -69,8 +70,10 @@ def _wait_for_admin_function(page: Page, function_name: str, timeout: int = 1000
         pytest.skip: If the function is not available within timeout
     """
     try:
-        page.wait_for_function(f"typeof window.Admin !== 'undefined' && typeof window.Admin.{function_name} === 'function'", timeout=timeout, polling=100)
-    except PlaywrightTimeoutError:
+        # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
+        # strict CSP (script-src 'self', no unsafe-eval) right after navigation.
+        wait_for_js_condition(page, f"typeof window.Admin !== 'undefined' && typeof window.Admin.{function_name} === 'function'", timeout=timeout, polling=100)
+    except TimeoutError:
         admin_debug = page.evaluate("""() => {
             return {
                 adminExists: typeof window.Admin !== 'undefined',
@@ -979,13 +982,16 @@ class TestAdminIframeContext:
         except PlaywrightTimeoutError:
             pass
 
-        # Wait for admin JS to initialise inside iframe
+        # Wait for admin JS to initialise inside iframe. evaluate()-based poll:
+        # wait_for_function's eval() mechanism is rejected by strict CSP
+        # (script-src 'self', no unsafe-eval) right after navigation.
         try:
-            frame_obj.wait_for_function(
+            wait_for_js_condition(
+                frame_obj,
                 "typeof window.Admin !== 'undefined' && typeof window.Admin.searchTeamSelector === 'function'",
                 timeout=15000,
             )
-        except PlaywrightTimeoutError:
+        except TimeoutError:
             pytest.fail("Admin JS did not initialise inside iframe - window.Admin.searchTeamSelector not available")
 
         frame = page.frame_locator("#admin-frame")
@@ -1011,9 +1017,11 @@ class TestAdminIframeContext:
             items_container = frame.locator("#team-selector-items")
             items_container.wait_for(state="visible", timeout=10000)
 
-            # Wait for actual team buttons (not "Loading..." placeholder)
-            frame_obj.wait_for_function(
-                "() => document.querySelectorAll('#team-selector-items .team-selector-item').length > 0",
+            # Wait for actual team buttons (not "Loading..." placeholder). evaluate()-based
+            # poll: wait_for_function's eval() mechanism is rejected by strict CSP.
+            wait_for_js_condition(
+                frame_obj,
+                "document.querySelectorAll('#team-selector-items .team-selector-item').length > 0",
                 timeout=15000,
             )
 

--- a/tests/playwright/test_admin_url_context.py
+++ b/tests/playwright/test_admin_url_context.py
@@ -73,7 +73,7 @@ def _wait_for_admin_function(page: Page, function_name: str, timeout: int = 1000
         # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
         # strict CSP (script-src 'self', no unsafe-eval) right after navigation.
         wait_for_js_condition(page, f"typeof window.Admin !== 'undefined' && typeof window.Admin.{function_name} === 'function'", timeout=timeout, polling=100)
-    except TimeoutError:
+    except PlaywrightTimeoutError:
         admin_debug = page.evaluate("""() => {
             return {
                 adminExists: typeof window.Admin !== 'undefined',
@@ -991,7 +991,7 @@ class TestAdminIframeContext:
                 "typeof window.Admin !== 'undefined' && typeof window.Admin.searchTeamSelector === 'function'",
                 timeout=15000,
             )
-        except TimeoutError:
+        except PlaywrightTimeoutError:
             pytest.fail("Admin JS did not initialise inside iframe - window.Admin.searchTeamSelector not available")
 
         frame = page.frame_locator("#admin-frame")

--- a/tests/playwright/test_admin_url_context.py
+++ b/tests/playwright/test_admin_url_context.py
@@ -70,8 +70,7 @@ def _wait_for_admin_function(page: Page, function_name: str, timeout: int = 1000
         pytest.skip: If the function is not available within timeout
     """
     try:
-        # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
-        # strict CSP (script-src 'self', no unsafe-eval) right after navigation.
+        # See wait_for_js_condition() docstring for why evaluate()-based polling is used.
         wait_for_js_condition(page, f"typeof window.Admin !== 'undefined' && typeof window.Admin.{function_name} === 'function'", timeout=timeout, polling=100)
     except PlaywrightTimeoutError:
         admin_debug = page.evaluate("""() => {
@@ -982,9 +981,7 @@ class TestAdminIframeContext:
         except PlaywrightTimeoutError:
             pass
 
-        # Wait for admin JS to initialise inside iframe. evaluate()-based poll:
-        # wait_for_function's eval() mechanism is rejected by strict CSP
-        # (script-src 'self', no unsafe-eval) right after navigation.
+        # Wait for admin JS to initialise inside iframe (see wait_for_js_condition() docstring).
         try:
             wait_for_js_condition(
                 frame_obj,
@@ -1017,8 +1014,8 @@ class TestAdminIframeContext:
             items_container = frame.locator("#team-selector-items")
             items_container.wait_for(state="visible", timeout=10000)
 
-            # Wait for actual team buttons (not "Loading..." placeholder). evaluate()-based
-            # poll: wait_for_function's eval() mechanism is rejected by strict CSP.
+            # Wait for actual team buttons (not "Loading..." placeholder); see
+            # wait_for_js_condition() docstring for why evaluate()-based polling is used.
             wait_for_js_condition(
                 frame_obj,
                 "document.querySelectorAll('#team-selector-items .team-selector-item').length > 0",

--- a/tests/playwright/test_innerhtml_guard.py
+++ b/tests/playwright/test_innerhtml_guard.py
@@ -15,6 +15,9 @@ correctly after converting from inline onclick to data-action + addEventListener
 import pytest
 from playwright.sync_api import expect
 
+# Local
+from .pages.admin_utils import wait_for_js_condition
+
 
 class TestToolTableButtons:
     """Tool table action buttons are loaded via fetch() + innerHTML.
@@ -197,6 +200,11 @@ class TestMetricsRetryButtons:
         page = admin_page.page
         admin_page.navigate()  # Ensure we're on admin page with JS loaded
 
+        # metrics.js is lazy-loaded on demand; window.Admin.showMetricsError only
+        # becomes callable once the metrics tab has triggered its chunk load.
+        page.locator("#tab-metrics").click()
+        wait_for_js_condition(page, "typeof window.Admin.showMetricsError === 'function'", timeout=15000)
+
         result = page.evaluate("""
             () => {
                 // Ensure required container exists
@@ -245,6 +253,11 @@ class TestMetricsRetryButtons:
         """displayMetrics() empty-state refresh button should have working click handler."""
         page = admin_page.page
         admin_page.navigate()  # Ensure we're on admin page with JS loaded
+
+        # metrics.js is lazy-loaded on demand; window.Admin.displayMetrics only
+        # becomes callable once the metrics tab has triggered its chunk load.
+        page.locator("#tab-metrics").click()
+        wait_for_js_condition(page, "typeof window.Admin.displayMetrics === 'function'", timeout=15000)
 
         result = page.evaluate("""
             () => {
@@ -870,6 +883,13 @@ class TestChatServerSelection:
             ),
         )
 
+        # llmChat.js is lazy-loaded on demand; window.Admin.loadFeature itself only
+        # becomes callable once a chunk that pulls it in has loaded, so trigger the
+        # load the same way a real user would (clicking the chat tab) rather than
+        # calling window.Admin.loadFeature directly.
+        page.locator('[data-testid="llm-chat-tab"]').click()
+        wait_for_js_condition(page, "typeof window.Admin.loadVirtualServersForChat === 'function'", timeout=15000)
+
         try:
             result = page.evaluate("""
                 async () => {
@@ -934,6 +954,13 @@ class TestChatServerSelection:
                 body='{"data": [{"id": "srv-a", "name": "Server A", "isActive": true, "visibility": "public", "description": "A", "associatedTools": []}, {"id": "srv-b", "name": "Server B", "isActive": true, "visibility": "public", "description": "B", "associatedTools": []}]}',
             ),
         )
+
+        # llmChat.js is lazy-loaded on demand; window.Admin.loadFeature itself only
+        # becomes callable once a chunk that pulls it in has loaded, so trigger the
+        # load the same way a real user would (clicking the chat tab) rather than
+        # calling window.Admin.loadFeature directly.
+        page.locator('[data-testid="llm-chat-tab"]').click()
+        wait_for_js_condition(page, "typeof window.Admin.loadVirtualServersForChat === 'function'", timeout=15000)
 
         try:
             result = page.evaluate("""

--- a/tests/playwright/test_organization.py
+++ b/tests/playwright/test_organization.py
@@ -14,6 +14,9 @@ import uuid
 import pytest
 from playwright.sync_api import expect
 
+# Local
+from .pages.admin_utils import wait_for_js_condition
+
 
 class TestTeams:
     """Tests for Team management features."""
@@ -124,9 +127,12 @@ class TestTeams:
         # Verify team is visible
         team_page.wait_for_team_visible(team_name)
 
-        # Wait for HTMX search to settle before interacting with cards
-        team_page.page.wait_for_function(
-            "() => !document.querySelector('#teams-loading.htmx-request')",
+        # Wait for HTMX search to settle before interacting with cards. evaluate()-based
+        # poll: wait_for_function's eval() mechanism is rejected by strict CSP
+        # (script-src 'self', no unsafe-eval) right after navigation/reload.
+        wait_for_js_condition(
+            team_page.page,
+            "!document.querySelector('#teams-loading.htmx-request')",
             timeout=15000,
         )
 
@@ -188,8 +194,11 @@ class TestTeams:
         team_search.fill(team_name)
         team_page.wait_for_team_visible(team_name)
 
-        team_page.page.wait_for_function(
-            "() => !document.querySelector('#teams-loading.htmx-request')",
+        # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
+        # strict CSP (script-src 'self', no unsafe-eval) right after navigation/reload.
+        wait_for_js_condition(
+            team_page.page,
+            "!document.querySelector('#teams-loading.htmx-request')",
             timeout=15000,
         )
 
@@ -296,9 +305,12 @@ class TestTeamSelectorDropdown:
         items_container = page.locator("#team-selector-items")
         expect(items_container).to_be_visible(timeout=10000)
 
-        # Wait for the loading message to be replaced with actual team buttons
-        page.wait_for_function(
-            "() => document.querySelectorAll('#team-selector-items .team-selector-item').length > 0",
+        # Wait for the loading message to be replaced with actual team buttons.
+        # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
+        # strict CSP (script-src 'self', no unsafe-eval) right after navigation.
+        wait_for_js_condition(
+            page,
+            "document.querySelectorAll('#team-selector-items .team-selector-item').length > 0",
             timeout=15000,
         )
 

--- a/tests/playwright/test_organization.py
+++ b/tests/playwright/test_organization.py
@@ -127,9 +127,8 @@ class TestTeams:
         # Verify team is visible
         team_page.wait_for_team_visible(team_name)
 
-        # Wait for HTMX search to settle before interacting with cards. evaluate()-based
-        # poll: wait_for_function's eval() mechanism is rejected by strict CSP
-        # (script-src 'self', no unsafe-eval) right after navigation/reload.
+        # Wait for HTMX search to settle before interacting with cards
+        # (see wait_for_js_condition() docstring for why evaluate()-based polling is used).
         wait_for_js_condition(
             team_page.page,
             "!document.querySelector('#teams-loading.htmx-request')",
@@ -194,8 +193,7 @@ class TestTeams:
         team_search.fill(team_name)
         team_page.wait_for_team_visible(team_name)
 
-        # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
-        # strict CSP (script-src 'self', no unsafe-eval) right after navigation/reload.
+        # See wait_for_js_condition() docstring for why evaluate()-based polling is used.
         wait_for_js_condition(
             team_page.page,
             "!document.querySelector('#teams-loading.htmx-request')",
@@ -305,9 +303,8 @@ class TestTeamSelectorDropdown:
         items_container = page.locator("#team-selector-items")
         expect(items_container).to_be_visible(timeout=10000)
 
-        # Wait for the loading message to be replaced with actual team buttons.
-        # evaluate()-based poll: wait_for_function's eval() mechanism is rejected by
-        # strict CSP (script-src 'self', no unsafe-eval) right after navigation.
+        # Wait for the loading message to be replaced with actual team buttons
+        # (see wait_for_js_condition() docstring for why evaluate()-based polling is used).
         wait_for_js_condition(
             page,
             "document.querySelectorAll('#team-selector-items .team-selector-item').length > 0",

